File: golang/binarytrees.go
package main

type TreeNodeObj struct {
	left  *TreeNodeObj
	right *TreeNodeObj
	item  int
}

func NewTreeNodeObj(item, depth int) *TreeNodeObj {
	node := &TreeNodeObj{item: item}
	if depth > 0 {
		shift := 1 << (depth - 1)
		node.left = NewTreeNodeObj(item-shift, depth-1)
		node.right = NewTreeNodeObj(item+shift, depth-1)
	}
	return node
}

func (t *TreeNodeObj) Sum() uint32 {
	total := uint32(t.item) + 1
	if t.left != nil {
		total += t.left.Sum()
	}
	if t.right != nil {
		total += t.right.Sum()
	}
	return total
}

type BinarytreesObj struct {
	BaseBenchmark
	n      int64
	result uint32
}

func (b *BinarytreesObj) Prepare() {
	b.n = b.ConfigVal("depth")
	b.result = 0
}

func (b *BinarytreesObj) Run(iteration_id int) {
	root := NewTreeNodeObj(0, int(b.n))
	b.result += root.Sum()

}

func (b *BinarytreesObj) Checksum() uint32 {
	return b.result
}

type TreeNodeArena struct {
	item  int
	left  int
	right int
}

type TreeArena struct {
	nodes []TreeNodeArena
}

func NewTreeArena() *TreeArena {
	return &TreeArena{
		nodes: make([]TreeNodeArena, 0),
	}
}

func (a *TreeArena) Build(item, depth int) int {
	idx := len(a.nodes)
	a.nodes = append(a.nodes, TreeNodeArena{item: item, left: -1, right: -1})

	if depth > 0 {
		shift := 1 << (depth - 1)
		leftIdx := a.Build(item-shift, depth-1)
		rightIdx := a.Build(item+shift, depth-1)
		a.nodes[idx].left = leftIdx
		a.nodes[idx].right = rightIdx
	}

	return idx
}

func (a *TreeArena) Sum(idx int) uint32 {
	node := a.nodes[idx]
	total := uint32(node.item) + 1

	if node.left >= 0 {
		total += a.Sum(node.left)
	}
	if node.right >= 0 {
		total += a.Sum(node.right)
	}

	return total
}

type BinarytreesArena struct {
	BaseBenchmark
	n      int64
	result uint32
}

func (b *BinarytreesArena) Prepare() {
	b.n = b.ConfigVal("depth")
	b.result = 0
}

func (b *BinarytreesArena) Run(iteration_id int) {
	arena := NewTreeArena()
	rootIdx := arena.Build(0, int(b.n))
	b.result += arena.Sum(rootIdx)
}

func (b *BinarytreesArena) Checksum() uint32 {
	return b.result
}

File: cpp/src/binarytrees.cpp
#include "binarytrees.hpp"

BinarytreesObj::TreeNode::TreeNode(int item, int depth) : item(item) {
  if (depth > 0) {
    left = std::make_unique<TreeNode>(item - (1 << (depth - 1)), depth - 1);
    right = std::make_unique<TreeNode>(item + (1 << (depth - 1)), depth - 1);
  }
}

BinarytreesObj::BinarytreesObj() : n(config_val("depth")), result_val(0) {}

std::string BinarytreesObj::name() const { return "Binarytrees::Obj"; }

void BinarytreesObj::run(int iteration_id) {
  (void)iteration_id;
  TreeNode root(0, n);
  result_val += root.sum();
}

uint32_t BinarytreesObj::checksum() { return result_val; }

BinarytreesArena::BinarytreesArena() : n(config_val("depth")), result_val(0) {}

std::string BinarytreesArena::name() const { return "Binarytrees::Arena"; }

int32_t BinarytreesArena::build_tree(int32_t item, int32_t depth) {
  int32_t idx = static_cast<int32_t>(arena.size());
  arena.emplace_back(item);

  if (depth > 0) {
    int32_t left_idx = build_tree(item - (1 << (depth - 1)), depth - 1);
    int32_t right_idx = build_tree(item + (1 << (depth - 1)), depth - 1);
    auto &node = arena[idx];
    node.left = left_idx;
    node.right = right_idx;
  }

  return idx;
}

uint32_t BinarytreesArena::sum(int32_t idx) const {
  const auto &node = arena[idx];
  uint32_t total = static_cast<uint32_t>(node.item) + 1;

  if (node.left >= 0) {
    total += sum(node.left);
  }
  if (node.right >= 0) {
    total += sum(node.right);
  }

  return total;
}

void BinarytreesArena::run(int iteration_id) {
  (void)iteration_id;
  arena = std::vector<TreeNode>();
  build_tree(0, static_cast<int32_t>(n));
  result_val += sum(0);
}

uint32_t BinarytreesArena::checksum() { return result_val; }